File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/ValueBar.java
package com.habitrpg.android.habitica.ui.views;

import android.content.Context;
import android.content.res.TypedArray;
import android.graphics.Bitmap;
import android.graphics.drawable.Drawable;
import android.support.v4.content.ContextCompat;
import android.util.AttributeSet;
import android.view.View;
import android.view.ViewGroup;
import android.widget.FrameLayout;
import android.widget.ImageView;
import android.widget.LinearLayout;
import android.widget.TextView;

import com.habitrpg.android.habitica.R;
import com.habitrpg.android.habitica.ui.helpers.DataBindingUtils;

import butterknife.BindView;
import butterknife.ButterKnife;

public class ValueBar extends FrameLayout {

    @BindView(R.id.ic_header)
    ImageView iconView;

    @BindView(R.id.valueLabel)
    TextView valueTextView;
    @BindView(R.id.descriptionLabel)
    TextView descriptionTextView;

    @BindView(R.id.bar_full)
    ViewGroup barBackground;
    @BindView(R.id.bar)
    View barView;
    @BindView(R.id.empty_bar_space)
    View barEmptySpace;


    public ValueBar(Context context, AttributeSet attrs) {
        super(context, attrs);

        inflate(context, R.layout.value_bar, this);
        ButterKnife.bind(this);

        TypedArray attributes = context.getTheme().obtainStyledAttributes(
                attrs,
                R.styleable.ValueBar,
                0, 0);
        setLightBackground(attributes.getBoolean(R.styleable.ValueBar_lightBackground, false));

        int color = attributes.getColor(R.styleable.ValueBar_barForegroundColor, 0);
        DataBindingUtils.INSTANCE.setRoundedBackground(barView, color);

        Drawable iconRes = attributes.getDrawable(R.styleable.ValueBar_barIconDrawable);
        if (iconRes != null) {
            setIcon(iconRes);
        }

        descriptionTextView.setText(attributes.getString(R.styleable.ValueBar_description));
    }

    public void setIcon(Drawable iconRes) {
        iconView.setImageDrawable(iconRes);
        iconView.setVisibility(View.VISIBLE);
    }

    public void setIcon(Bitmap bitmap) {
        iconView.setImageBitmap(bitmap);
        iconView.setVisibility(View.VISIBLE);
    }

    public void setBarWeight(double percent) {
        setLayoutWeight(barView, percent);
        setLayoutWeight(barEmptySpace, 1.0f - percent);
    }

    public void setValueText(String valueText) {
        valueTextView.setText(valueText);
    }

    public void setLightBackground(boolean lightBackground) {
        int textColor;
        if (lightBackground) {
            textColor = ContextCompat.getColor(getContext(), R.color.gray_10);
            barBackground.setBackgroundResource(R.drawable.layout_rounded_bg_light_gray);
        } else {
            textColor = ContextCompat.getColor(getContext(), R.color.brand_500);
            barBackground.setBackgroundResource(R.drawable.layout_rounded_bg_brand);
        }
        valueTextView.setTextColor(textColor);
        descriptionTextView.setTextColor(textColor);
    }

    public static void setLayoutWeight(View view, double weight) {
        view.clearAnimation();
        LinearLayout.LayoutParams layout = (LinearLayout.LayoutParams) view.getLayoutParams();
        if (weight == 0.0f || weight == 1.0f) {
            layout.weight = (float) weight;
            view.setLayoutParams(layout);
        } else if (layout.weight != weight) {
            DataBindingUtils.LayoutWeightAnimation anim = new DataBindingUtils.LayoutWeightAnimation(view, (float) weight);
            anim.setDuration(1250);
            view.startAnimation(anim);
        }
    }

    public void set(double value, double valueMax) {
        double percent = Math.min(1, value / valueMax);

        this.setBarWeight(percent);
        this.setValueText((int) value + "/" + (int) valueMax);
    }
}

File: Habitica/src/main/java/com/habitrpg/android/habitica/ui/views/ValueBar.kt
package com.habitrpg.android.habitica.ui.views

import android.content.Context
import android.graphics.Bitmap
import android.graphics.drawable.Drawable
import android.util.AttributeSet
import android.view.View
import android.view.ViewGroup
import android.widget.FrameLayout
import android.widget.ImageView
import android.widget.TextView
import androidx.core.content.ContextCompat
import com.habitrpg.android.habitica.R
import com.habitrpg.android.habitica.extensions.getThemeColor
import com.habitrpg.android.habitica.ui.helpers.bindView
import java.math.RoundingMode
import java.text.NumberFormat


class ValueBar(context: Context, attrs: AttributeSet?) : FrameLayout(context, attrs) {

    private val iconView: ImageView by bindView(R.id.ic_header)
    private val secondaryIconView: ImageView by bindView(R.id.secondaryIconView)
    private val descriptionIconView: ImageView by bindView(R.id.descriptionIconView)
    private val valueTextView: TextView by bindView(R.id.valueLabel)
    private val descriptionTextView: TextView by bindView(R.id.descriptionLabel)
    private val progressBar: HabiticaProgressBar by bindView(R.id.progressBar)
    private val labelWrapper: ViewGroup by bindView(R.id.labelWrapper)

    private val formatter = NumberFormat.getInstance()

    private var currentValue: Double = 0.0
    set(value) {
        field = value
        updateBar()
    }

    var pendingValue: Double = 0.0
        set(value) {
            field = value
            updateBar()
        }

    private var maxValue: Double = 0.0
        set(value) {
            field = value
            updateBar()
        }

    var barHeight: Int? = null
    set(value) {
        field = value
        if (value != null) {
            progressBar.layoutParams.height = value
        }
    }

    var description: String = ""
    set(value) {
        field = value
        descriptionTextView.text = description
    }

    private fun updateBar() {
        this.progressBar.pendingValue = pendingValue
        this.progressBar.currentValue = currentValue
        this.progressBar.maxValue = maxValue
        this.setValueText(formatter.format(currentValue) + " / " + formatter.format(maxValue.toInt()))
    }

    init {
        View.inflate(context, R.layout.value_bar, this)

        val attributes = context.theme?.obtainStyledAttributes(
                attrs,
                R.styleable.ValueBar,
                0, 0)
        setLightBackground(attributes?.getBoolean(R.styleable.ValueBar_lightBackground, false) == true)

        progressBar.barForegroundColor = attributes?.getColor(R.styleable.ValueBar_barForegroundColor, 0) ?: 0
        progressBar.barPendingColor = attributes?.getColor(R.styleable.ValueBar_barPendingColor, 0) ?: 0
        progressBar.barBackgroundColor = attributes?.getColor(R.styleable.ValueBar_barBackgroundColor, 0) ?: 0

        val labelSpacing = attributes?.getDimension(R.styleable.ValueBar_labelSpacing, context.resources.getDimension(R.dimen.spacing_small))
        if (labelSpacing != null) {
            labelWrapper.setPadding(0, labelSpacing.toInt(), 0, 0)
        }

        barHeight = attributes?.getDimension(R.styleable.ValueBar_barHeight, context.resources.getDimension(R.dimen.bar_size))?.toInt()

        val valueTextColor = attributes?.getColor(R.styleable.ValueBar_valueTextColor, 0) ?: 0
        if (valueTextColor != 0) {
            valueTextView.setTextColor(valueTextColor)
        }
        val descriptionTextColor = attributes?.getColor(R.styleable.ValueBar_descriptionTextColor, 0) ?: 0
        if (descriptionTextColor != 0) {
            descriptionTextView.setTextColor(descriptionTextColor)
        }

        val iconRes = attributes?.getDrawable(R.styleable.ValueBar_barIconDrawable)
        if (iconRes != null) {
            setIcon(iconRes)
        }

        descriptionTextView.text = attributes?.getString(R.styleable.ValueBar_description)

        formatter.maximumFractionDigits = 1
        formatter.roundingMode = RoundingMode.UP
        formatter.isGroupingUsed = true
    }

    fun setIcon(iconRes: Drawable) {
        iconView.setImageDrawable(iconRes)
        iconView.visibility = View.VISIBLE
    }

    fun setIcon(bitmap: Bitmap) {
        iconView.setImageBitmap(bitmap)
        iconView.visibility = View.VISIBLE
    }

    fun setSecondaryIcon(iconRes: Drawable) {
        secondaryIconView.setImageDrawable(iconRes)
        secondaryIconView.visibility = View.VISIBLE
    }

    fun setSecondaryIcon(bitmap: Bitmap) {
        secondaryIconView.setImageBitmap(bitmap)
        secondaryIconView.visibility = View.VISIBLE
    }

    fun setDescriptionIcon(iconRes: Drawable) {
        descriptionIconView.setImageDrawable(iconRes)
        descriptionIconView.visibility = View.VISIBLE
    }

    fun setDescriptionIcon(bitmap: Bitmap) {
        descriptionIconView.setImageBitmap(bitmap)
        descriptionIconView.visibility = View.VISIBLE
    }

    fun setValueText(valueText: String) {
        valueTextView.text = valueText
    }

    fun setLightBackground(lightBackground: Boolean) {
        val textColor: Int
        if (lightBackground) {
            textColor = ContextCompat.getColor(context, R.color.gray_10)
            progressBar.setBackgroundResource(R.drawable.layout_rounded_bg_light_gray)
        } else {
            textColor = context.getThemeColor(R.attr.textColorPrimaryDark)
            progressBar.setBackgroundResource(R.drawable.layout_rounded_bg_primary)
        }
        valueTextView.setTextColor(textColor)
        descriptionTextView.setTextColor(textColor)
    }

    fun set(value: Double, valueMax: Double) {
        currentValue = value
        maxValue = valueMax
    }

    fun setLabelVisibility(visibility: Int) {
        valueTextView.visibility = visibility
        descriptionTextView.visibility = visibility
    }
}
